File: GameServer/internal/parser.go
package internal

import (
	"fmt"
	"gameserver/internal/logger"
	"gameserver/internal/models"
	"gameserver/internal/utils"
	"gameserver/internal/utils/errorHandeling"
	"strconv"
	"strings"
	"unsafe"
) //todo predelat bez pouziti regex

// region CONSTANTS
/*var gRegexPatterns = map[string]string{
	"parseGameIDPlayerID":  `^\{"gameID":"(\d+)","playerID":"(\d+)"\}$`,
	"parsePlayerLoginArgs": `^\{"nickname":"([A-Za-z0-9_\\-]+)"\}$`,
	"parseCreateGameArgs":  `^\{"name":"([A-Za-z0-9_\\-]+)","maxPlayers":"(\d+)"\}$`,
	"parsePlayerID":        `^\{"playerID":"(\d+)"\}$`,
}*/

//endregion

// region DATA STRUCTURES
type Brackets struct {
	Opening string
	Closing string
}

var (
	cParamsBrackets          = Brackets{Opening: "{", Closing: "}"}
	cArrayBrackets           = Brackets{Opening: "[", Closing: "]"}
	cParamsDelimiter         = ","
	cParamsKeyValueDelimiter = ":"
	cParamsWrapper           = "\""
)

//endregion

// region FUNCTIONS PARSE
func ParseMessage(input string) (models.Message, error) {
	logger.Log.Debug("Parsing message: %v", input)

	//remove end delimiter
	if len(input) > 0 && input[len(input)-1] == utils.CMessageEndDelimiter {
		input = input[:len(input)-1]
	}

	messageData := models.MessageHeader{}
	signatureSize := int(unsafe.Sizeof(messageData.Signature))
	commandIDSize := int(unsafe.Sizeof(messageData.CommandID))
	timeStempSize := int(unsafe.Sizeof(messageData.TimeStamp))
	minSizeMessage := signatureSize + commandIDSize + timeStempSize

	if len(input) < minSizeMessage {
		return models.Message{}, fmt.Errorf("invalid input format")
	}

	start := 0

	//Read Signature
	signature := input[start : start+signatureSize]
	start += signatureSize

	//Read Command ID
	commandID := input[start : start+commandIDSize]
	start += commandIDSize

	//Read TimeStemp
	timeStamp := input[start : start+timeStempSize]
	start += timeStempSize

	//Read nickname
	playerNickname, playerNicknameSize, err := parseMessagePlayerID(input[start:])
	if err != nil {
		errorHandeling.PrintError(err)
		return models.Message{}, fmt.Errorf("error parsing player ID: %v", err)
	}
	start += playerNicknameSize

	//Read Parameters
	parametersStr := input[start:]

	params, err := parseParamsStr(parametersStr)
	if err != nil {
		errorHandeling.PrintError(err)
		return models.Message{}, fmt.Errorf("error parsing params: %v", err)
	}

	//convert values
	commandIDint, err := strconv.ParseUint(commandID, 10, commandIDSize)
	if err != nil {
		errorHandeling.PrintError(err)
		return models.Message{}, fmt.Errorf("error parsing command ID: %v", err)
	}

	return models.Message{
		Signature:      signature, // You can populate this with the actual signature values
		CommandID:      int(commandIDint),
		TimeStamp:      timeStamp,
		PlayerNickname: playerNickname,
		Parameters:     params,
	}, nil
}

func ParseParamValueArray(value string) ([]string, error) {
	var valueArray []string

	if len(value) == 0 {
		return valueArray, nil
	}

	if value[0] != cArrayBrackets.Opening[0] {
		return valueArray, fmt.Errorf("invalid valueArray format")
	}

	if value[len(value)-1] != cArrayBrackets.Closing[0] {
		return valueArray, fmt.Errorf("invalid valueArray format")
	}

	value = value[1 : len(value)-1]

	valuesStr := strings.Split(value, cParamsDelimiter)

	for _, valueStr := range valuesStr {
		valueArray = append(valueArray, valueStr)
	}

	return valueArray, nil
}

func parseParamsStr(paramsString string) ([]utils.Params, error) {
	var paramArray []utils.Params

	if len(paramsString) == 0 {
		return paramArray, nil
	}

	if paramsString[0] != cParamsBrackets.Opening[0] {
		return paramArray, fmt.Errorf("invalid paramArray format")
	}

	if paramsString[len(paramsString)-1] != cParamsBrackets.Closing[0] {
		return paramArray, fmt.Errorf("invalid paramArray format")
	}

	paramsString = paramsString[1 : len(paramsString)-1]

	paramsStr := strings.Split(paramsString, cParamsDelimiter)

	for _, paramStr := range paramsStr {
		parameter, err := parseParam(paramStr)
		if err != nil {
			errorHandeling.PrintError(err)
			return paramArray, fmt.Errorf("invalid paramArray format")
		}
		paramArray = append(paramArray, parameter)
	}

	return paramArray, nil
}

func parseParam(str string) (utils.Params, error) {
	parameter := utils.Params{}

	if len(str) == 0 {
		return parameter, nil
	}

	paramKeyValue := strings.Split(str, cParamsDelimiter)

	if len(paramKeyValue) != 2 {
		return parameter, fmt.Errorf("invalid param format")
	}

	parameter.Name = paramKeyValue[0]
	if parameter.Name[0] != cParamsWrapper[0] {
		return parameter, fmt.Errorf("invalid param format")
	}

	if parameter.Name[len(parameter.Name)-1] != cParamsWrapper[0] {
		return parameter, fmt.Errorf("invalid param format")
	}
	parameter.Name = parameter.Name[1 : len(parameter.Name)-1]

	parameter.Value = paramKeyValue[1]
	if parameter.Value[0] != cParamsWrapper[0] {
		return parameter, fmt.Errorf("invalid param format")
	}
	if parameter.Value[len(parameter.Value)-1] != cParamsWrapper[0] {
		return parameter, fmt.Errorf("invalid param format")
	}
	parameter.Value = parameter.Value[1 : len(parameter.Value)-1]

	return parameter, nil
}

func parseMessagePlayerID(input string) (string, int, error) {
	playerNickname := ""
	playerNicknameSize := 0

	if len(input) == 0 {
		return playerNickname, playerNicknameSize, fmt.Errorf("invalid player ID format")
	}

	opening := cParamsBrackets.Opening
	closing := cParamsBrackets.Closing

	playerNickname, err := extractSubstring(input, opening, closing)
	if err != nil {
		errorHandeling.PrintError(err)
		return playerNickname, playerNicknameSize, fmt.Errorf("invalid player ID format")
	}

	playerNicknameSize = len(playerNickname) + len(opening) + len(closing)

	return playerNickname, playerNicknameSize, nil
}

//endregion

// region FUNCTIONS UTILS
func extractSubstring(input, opening, closing string) (string, error) {
	startIndex := strings.Index(input, opening)
	if startIndex == -1 {
		return "", fmt.Errorf("opening character not found")
	}
	endIndex := strings.Index(input[startIndex:], closing)
	if endIndex == -1 {
		return "", fmt.Errorf("closing character not found")
	}
	return input[startIndex+len(opening) : startIndex+endIndex], nil
}

func getArrayElementStr(name string, value string) string {
	returnStr := cParamsWrapper + name + cParamsWrapper + cParamsKeyValueDelimiter + cParamsWrapper + value + cParamsWrapper + cParamsDelimiter
	return returnStr
}

//endregion

// region FUNCTIONS CONVERT
func ConvertMessageToNetworkString(message models.Message) (string, error) {
	networkString := ""

	messageData := models.MessageHeader{}
	signatureSize := int(unsafe.Sizeof(messageData.Signature))
	commandIDSize := int(unsafe.Sizeof(messageData.CommandID))
	timeStempSize := int(unsafe.Sizeof(messageData.TimeStamp))

	//Signature
	if len(message.Signature) != signatureSize {
		return networkString, fmt.Errorf("invalid signature")
	}
	networkString += message.Signature

	//Command ID
	commandIDStr := fmt.Sprintf("%02d", message.CommandID)
	if len(commandIDStr) != commandIDSize {
		return networkString, fmt.Errorf("invalid command ID")
	}
	networkString += commandIDStr

	//TimeStemp
	if len(message.TimeStamp) != timeStempSize {
		return networkString, fmt.Errorf("invalid time stamp")
	}
	networkString += message.TimeStamp

	//PlayerNickname
	networkString += convertPlayerNicknameToNetworkString(message.PlayerNickname)

	//Parameters
	paramsStr, err := convertParamsToNetworkString(message.Parameters)
	if err != nil {
		errorHandeling.PrintError(err)
		return networkString, err
	}
	networkString += paramsStr

	return networkString, nil
}

func ConvertGameListToNetworkString(array []*models.Game) string {
	arrayStr := ""

	arrayStr += cArrayBrackets.Opening

	for _, game := range array {

		gameName := game.GetName()
		maxPlayers := fmt.Sprintf("%d", game.GetMaxPlayers())
		playersCount := fmt.Sprintf("%d", game.GetPlayersCount())

		arrayStr += cParamsBrackets.Opening
		arrayStr += getArrayElementStr("name", gameName)
		arrayStr += getArrayElementStr("maxPlayers", maxPlayers)
		arrayStr += getArrayElementStr("playersCount", playersCount)
		arrayStr += cParamsBrackets.Closing + cParamsDelimiter
	}

	arrayStr += cArrayBrackets.Closing

	return arrayStr
}

func ConvertPlayerListToNetworkString(array []*models.Player) string {
	arrayStr := ""

	arrayStr += cArrayBrackets.Opening

	for _, player := range array {

		playerNickname := player.GetNickname()

		arrayStr += cParamsBrackets.Opening
		arrayStr += cParamsWrapper + "nickname" + cParamsWrapper + cParamsKeyValueDelimiter + cParamsWrapper + playerNickname + cParamsWrapper
		arrayStr += cParamsBrackets.Closing + cParamsDelimiter
	}

	arrayStr += cArrayBrackets.Closing

	return arrayStr
}

func ConvertGameDataToNetworkString(data models.GameData) string {
	arrayStr := ""

	arrayStr += cArrayBrackets.Opening

	for _, playerGameData := range data.PlayerGameDataArr {

		nickname := playerGameData.Player.GetNickname()
		isConnected := fmt.Sprintf("%t", playerGameData.Player.IsConnected())
		score := fmt.Sprintf("%d", playerGameData.Score)
		isTurnBool := playerGameData.Player == data.TurnPlayer
		isTurn := fmt.Sprintf("%t", isTurnBool)

		arrayStr += cParamsBrackets.Opening
		arrayStr += getArrayElementStr("nickname", nickname)
		arrayStr += getArrayElementStr("isConnected", isConnected)
		arrayStr += getArrayElementStr("score", score)
		arrayStr += getArrayElementStr("isTurn", isTurn)

		arrayStr += cParamsBrackets.Closing + cParamsDelimiter
	}

	arrayStr += cArrayBrackets.Closing

	return arrayStr
}

func ConvertCubeValuesToNetworkString(values []int) string {
	arrayStr := ""

	arrayStr += cArrayBrackets.Opening

	for _, value := range values {

		valueStr := fmt.Sprintf("%d", value)

		arrayStr += cParamsBrackets.Opening
		arrayStr += getArrayElementStr("value", valueStr)
		arrayStr += cParamsBrackets.Closing + cParamsDelimiter
	}

	arrayStr += cArrayBrackets.Closing

	return arrayStr
}

func convertPlayerNicknameToNetworkString(nickname string) string {
	networkStr := ""
	networkStr += cParamsBrackets.Opening
	networkStr += nickname
	networkStr += cParamsBrackets.Closing

	return networkStr
}

func convertParamsToNetworkString(params []utils.Params) (string, error) {
	length := len(params)

	paramsStr := ""
	paramsStr += cParamsBrackets.Opening
	for i := 0; i < length; i++ {
		paramsStr += cParamsWrapper + params[i].Name + cParamsWrapper + cParamsKeyValueDelimiter + cParamsWrapper + params[i].Value + cParamsWrapper + cParamsDelimiter
	}
	paramsStr += cParamsBrackets.Closing

	return paramsStr, nil
}

//endregion

File: GameClient/src/parser.py
from typing import List

from GUI import Game

TIMESTAMP_SIZE = 32

COMMANDID_SIZE = 2

SIGNATURE_SIZE = 6


class Brackets:
    def __init__(self, opening, closing):
        self.opening = opening
        self.closing = closing


c_params_brackets = Brackets("{", "}")
c_array_brackets = Brackets("[", "]")
c_params_delimiter = ","
c_params_key_value_delimiter = ":"
c_params_wrapper = "\""



class Param:
    def __init__(self, name: str, value: str):
        self.name = name
        self.value = value

class NetworkMessage:
    def __init__(self, signature: str, command_id: int, timestamp: str, player_nickname: str, parameters: List[Param]):
        self.signature = signature
        self.command_id = command_id
        self.timestamp = timestamp
        self.player_nickname = player_nickname
        self.parameters = parameters



def convert_params_to_network_string(params: List[Param]) -> str:
    params_str = c_params_brackets.opening
    for param in params:
        params_str += c_params_wrapper + param.name + c_params_wrapper + c_params_key_value_delimiter + c_params_wrapper + param.value + c_params_wrapper + c_params_delimiter
    params_str = params_str.rstrip(c_params_delimiter)  # remove the trailing delimiter
    params_str += c_params_brackets.closing

    return params_str

def parse_message(input_str):
    signature_size = SIGNATURE_SIZE
    command_id_size = COMMANDID_SIZE
    timestamp_size = TIMESTAMP_SIZE
    min_size_message = signature_size + command_id_size + timestamp_size

    if len(input_str) < min_size_message:
        raise ValueError("Invalid input format")

    start = 0

    # Read Signature
    signature = input_str[start : start + signature_size]
    start += signature_size
    start += 1

    # Read Command ID
    command_id = input_str[start : start + command_id_size]
    start += command_id_size
    start += 1

    # Read Timestamp
    timestamp = input_str[start : start + timestamp_size]
    start += timestamp_size
    start += 1

    # Read nickname
    player_nickname, player_nickname_size = parse_message_player_id(input_str[start:])
    start += player_nickname_size
    start += 1

    # Read Parameters
    parameters_str = input_str[start:]

    params = parse_params_str(parameters_str)

    # Convert values
    command_id_int = int(command_id)

    return NetworkMessage(signature, command_id_int, timestamp, player_nickname, params)


def parse_params_str(params_string):
    param_array = []

    if len(params_string) == 0:
        return param_array

    if params_string[0] != c_params_brackets.opening:
        raise ValueError("Invalid paramArray format")

    if params_string[-1] != c_params_brackets.closing:
        raise ValueError("Invalid paramArray format")

    params_string = params_string[1:-1]

    params_str = params_string.split(c_params_delimiter)

    for param_str in params_str:
        parameter = parse_param(param_str)
        param_array.append(parameter)

    return param_array


def parse_param(str):
    if len(str) == 0:
        return Param("", "")

    param_key_value = str.split(c_params_key_value_delimiter)

    if len(param_key_value) != 2:
        raise ValueError("Invalid param format")

    name = param_key_value[0]
    if name[0] != c_params_wrapper or name[-1] != c_params_wrapper:
        raise ValueError("Invalid param format")
    name = name[1:-1]

    value = param_key_value[1]
    if value[0] != c_params_wrapper or value[-1] != c_params_wrapper:
        raise ValueError("Invalid param format")
    value = value[1:-1]

    return Param(name, value)


def parse_message_player_id(input):
    player_nickname = ""
    player_nickname_size = 0

    if len(input) == 0:
        raise ValueError("Invalid player ID format")

    opening = c_params_brackets.opening
    closing = c_params_brackets.closing

    player_nickname = input[len(opening):-len(closing)]
    player_nickname_size = len(player_nickname) + len(opening) + len(closing)

    return player_nickname, player_nickname_size

def convert_message_to_network_string(message: NetworkMessage) -> str:
    network_string = ""

    signature_size = SIGNATURE_SIZE
    command_id_size = COMMANDID_SIZE
    timestamp_size = TIMESTAMP_SIZE

    # Signature
    if len(message.signature) != signature_size:
        raise ValueError("Invalid signature")
    network_string += message.signature

    # Command ID
    command_id_str = str(message.command_id)
    if len(command_id_str) != command_id_size:
        raise ValueError("Invalid command ID")
    network_string += command_id_str

    # Timestamp
    if len(message.timestamp) != timestamp_size:
        raise ValueError("Invalid timestamp")
    network_string += message.timestamp

    # PlayerNickname
    network_string += c_params_brackets.opening
    network_string += message.player_nickname
    network_string += c_params_brackets.closing

    # Parameters
    params_str = convert_params_to_network_string(message.parameters)
    network_string += params_str

    return network_string


def parse_game(game_str):
    game = Game("", 0)
    if len(game_str) == 0:
        return game

    if game_str[0] != c_params_brackets.opening:
        raise ValueError("Invalid game format")

    if game_str[-1] != c_params_brackets.closing:
        raise ValueError("Invalid game format")

    game_str = game_str[1:-1]

    game_str = game_str.split(c_params_key_value_delimiter)

    if len(game_str) != 2:
        raise ValueError("Invalid game format")

    name = game_str[0]
    if name[0] != c_params_wrapper or name[-1] != c_params_wrapper:
        raise ValueError("Invalid game format")
    name = name[1:-1]

    max_players = game_str[1]
    if max_players[0] != c_params_wrapper or max_players[-1] != c_params_wrapper:
        raise ValueError("Invalid game format")
    max_players = max_players[1:-1]

    player_count game_str[2]
    if player_count[0] != c_params_wrapper or player_count[-1] != c_params_wrapper:
        raise ValueError("Invalid game format")

    game = Game(name, max_players, player_count)
    #todo check if wrong

    return game


def convert_params_game_list(parameters: List[Param]) -> List[Game]:
    # Convert the parameters to a list of Game objects
    games = []
    if len(parameters) != 0:
        return games
    param_value = parameters[0].value

    if param_value[0] != c_array_brackets.opening or param_value[-1] != c_array_brackets.closing:
        raise ValueError("Invalid param value format")
    param_value = param_value[1:-1]

    param_value = param_value.split(c_params_delimiter)
    for game_str in param_value:

        game = parse_game(game_str)
        games.append(game)

    return games